eta/core/datasets.py

```diff
@@ -1338,3 +1373,839 @@ def required(cls):
+            # object indices for objects of type self.object_label.
+            # The values are unique attribute values for self.attr_name.
+
+            if len(helper_dict):
+                helper = (i, [])
+                for s in helper_dict.values():
+                    helper[1].extend(s)
+                helper_list.append(helper)
+
+        return helper_list
+
+    def _get_target_count(self, counts):
+        '''Compute the target count that we would like to balance each value to.
+
+        Args:
+            A: the occurrence matrix computed in Balancer._get_occurrence_matrix
+
+        Returns: Integer target value
+        '''
+        if self.target_count:
+            return self.target_count
+        return int(np.quantile(counts, self.target_quantile))
+
+    def _get_keep_idxs(self, A, counts, target_count):
+        '''Algorithm fun! This function chooses the set of records to keep (and
+        remove).
+
+        There's still plenty of potential for testing and improvement here.
+
+
+        This problem can be posed as:
+            minimize:
+                ||np.dot(A, x) - b||
+            subject to:
+                x[i] is an element of [0, 1]
+
+        and different algorithms may be substituted in.
+
+        Args:
+            A: the occurrence matrix computed in Balancer._get_occurrence_matrix
+            counts: (vector) the original counts for each value
+            target_count: (int) the target value
+
+        Returns:
+            a list of integer indices to keep
+        '''
+        b = counts - target_count
+
+        x = self._greedy(A, b)
+
+        return np.where(x==0)[0]
+
+    def _random(self, A, b):
+        '''A random search algorithm for finding the indices to omit.
+
+        Args:
+            A: the occurrence matrix
+            b: the target vector to match
+
+        Returns:
+            x: the solution vector. x[j]=1 -> omit the j'th record
+        '''
+        best_x = np.zeros(A.shape[1], dtype=np.dtype('int'))
+        best_score = self._solution_score(b - np.dot(A, best_x))
+
+        import random
+        random.seed(1)
+        for _ in range(10000):
+            i = random.choice(np.where(best_x == 0)[0])
+            cur_x = best_x.copy()
+            cur_x[i] = 1
+            cur_score = self._solution_score(b - np.dot(A, cur_x))
+            if cur_score < best_score:
+                best_score = cur_score
+                best_x = cur_x
+
+        return best_x
+
+    def _greedy(self, A, b):
+        '''A greedy search algorithm for finding the indices to omit.
+
+        Args:
+            A: the occurrence matrix
+            b: the target vector to match
+
+        Returns:
+            x: the solution vector. x[j]=1 -> omit the j'th record
+        '''
+        best_x = np.zeros(A.shape[1], dtype=np.dtype('int'))
+        best_score = self._solution_score(b - np.dot(A, best_x))
+
+        while True:
```

Comment:
I haven't translated to math to understand what you're doing, but `while True` scares me

Reply:
haha, yeah it's just a do-while loop. I changed the code, but the main condition is still used with a `if <this>: break`